packages/snaps-controllers/src/snaps/SnapController.ts

```diff
@@ -1548,43 +1573,76 @@ export class SnapController extends BaseController<
   ): Promise<InstallSnapsResult> {
     const result: InstallSnapsResult = {};
 
-    await Promise.all(
-      Object.entries(requestedSnaps).map(
-        async ([snapId, { version: rawVersion }]) => {
-          const [error, version] = resolveVersionRange(rawVersion);
-          if (error) {
-            result[snapId] = {
-              error: ethErrors.rpc.invalidParams(
-                `The "version" field must be a valid SemVer version range if specified. Received: "${version}".`,
-              ),
-            };
-            return;
-          }
-          const permissionName = getSnapPermissionName(snapId);
+    const snapIds = Object.keys(requestedSnaps);
 
-          if (
-            this.messagingSystem.call(
-              'PermissionController:hasPermission',
-              origin,
-              permissionName,
-            )
-          ) {
-            // Attempt to install and run the snap, storing any errors that
-            // occur during the process.
-            result[snapId] = {
-              ...(await this.processRequestedSnap(origin, snapId, version)),
-            };
-          } else {
-            // only allow the installation of permitted snaps
-            result[snapId] = {
-              error: ethErrors.provider.unauthorized(
-                `Not authorized to install snap "${snapId}". Request the permission for the snap before attempting to install it.`,
-              ),
-            };
-          }
-        },
-      ),
+    // Existing snaps may need to be updated
+    const pendingUpdates = snapIds.filter((snapId) => this.has(snapId));
+
+    // Non-existing snaps will need to be installed
+    const pendingInstalls = snapIds.filter(
+      (snapId) => !pendingUpdates.includes(snapId),
     );
+
+    try {
+      await Promise.all(
+        Object.entries(requestedSnaps).map(
+          async ([snapId, { version: rawVersion }]) => {
+            const [error, version] = resolveVersionRange(rawVersion);
+
+            if (error) {
+              throw ethErrors.rpc.invalidParams(
+                `The "version" field must be a valid SemVer version range if specified. Received: "${rawVersion}".`,
+              );
+            }
+
+            const permissionName = getSnapPermissionName(snapId);
+
+            if (
+              !this.messagingSystem.call(
+                'PermissionController:hasPermission',
+                origin,
+                permissionName,
+              )
+            ) {
+              throw ethErrors.provider.unauthorized(
+                `Not authorized to install snap "${snapId}". Request the permission for the snap before attempting to install it.`,
+              );
+            }
+
+            const isUpdate = pendingUpdates.includes(snapId);
```

Comment:
Right now this will create snapshots for any installed snaps, even if they don't end up needing an update. Not sure if we want to detect that up front or not.

Reply:
This complicates things a bit more, we don't truly know a snapshot is not warranted until the `gtVersion` call inside `updateSnap` which happens after we have a resolved version from the `fetchSnap` call. Going to have to figure out a way to either prevent this scenario, or allow for an empty snapshot and be able to check for that inside tests.

Reply:
Figured it out.